## Week 4, Lab 2: Predicting Chronic Kidney Disease in Patients
> Author: Matt Brems

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

In this lab, we're going to focus on steps exploring data, building models and evaluating the models we build.

There are three links you may find important:
- [A set of chronic kidney disease (CKD) data and other biological factors](./chronic_kidney_disease_full.csv).
- [The CKD data dictionary](./chronic_kidney_disease_header.txt).
- [An article comparing the use of k-nearest neighbors and support vector machines on predicting CKD](./chronic_kidney_disease.pdf).

## Step 1: Define the problem.

Suppose you're working for Mayo Clinic, widely recognized to be the top hospital in the United States. In your work, you've overheard nurses and doctors discuss test results, then arrive at a conclusion as to whether or not someone has developed a particular disease or condition. For example, you might overhear something like:

> **Nurse**: Male 57 year-old patient presents with severe chest pain. FDP _(short for fibrin degradation product)_ was elevated at 13. We did an echo _(echocardiogram)_ and it was inconclusive.

> **Doctor**: What was his interarm BP? _(blood pressure)_

> **Nurse**: Systolic was 140 on the right; 110 on the left.

> **Doctor**: Dammit, it's an aortic dissection! Get to the OR _(operating room)_ now!

> _(intense music playing)_

In this fictitious but [Shonda Rhimes-esque](https://en.wikipedia.org/wiki/Shonda_Rhimes#Grey's_Anatomy,_Private_Practice,_Scandal_and_other_projects_with_ABC) scenario, you might imagine the doctor going through a series of steps like a [flowchart](https://en.wikipedia.org/wiki/Flowchart), or a series of if-this-then-that steps to diagnose a patient. The first steps made the doctor ask what the interarm blood pressure was. Because interarm blood pressure took on the values it took on, the doctor diagnosed the patient with an aortic dissection.

Your goal, as a research biostatistical data scientist at the nation's top hospital, is to develop a medical test that can improve upon our current diagnosis system for [chronic kidney disease (CKD)](https://www.mayoclinic.org/diseases-conditions/chronic-kidney-disease/symptoms-causes/syc-20354521).

**Real-world problem**: Develop a medical diagnosis test that is better than our current diagnosis system for CKD.

**Data science problem**: Develop a medical diagnosis test that reduces both the number of false positives and the number of false negatives.

---

## Step 2: Obtain the data.

### 1. Read in the data.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('chronic_kidney_disease_full.csv')

### 2. Check out the data dictionary. What are a few features or relationships you might be interested in checking out?

Answer: 

* 2. Blood Pressure
* 11. Blood Urea
* 12. serum Creatinine
* 13. Sodium(numerical)
* 19. Hypertension

---

## Step 3: Explore the data.

### 3. How much of the data is missing from each column?

In [3]:
pd.options.display.max_rows = 99
pd.options.display.max_columns = 99

In [4]:
df.shape

(400, 25)

In [5]:
df.head(400)

age     bp     sg   al   su       rbc        pc         pcc          ba  \
0    48.0   80.0  1.020  1.0  0.0       NaN    normal  notpresent  notpresent   
1     7.0   50.0  1.020  4.0  0.0       NaN    normal  notpresent  notpresent   
2    62.0   80.0  1.010  2.0  3.0    normal    normal  notpresent  notpresent   
3    48.0   70.0  1.005  4.0  0.0    normal  abnormal     present  notpresent   
4    51.0   80.0  1.010  2.0  0.0    normal    normal  notpresent  notpresent   
5    60.0   90.0  1.015  3.0  0.0       NaN       NaN  notpresent  notpresent   
6    68.0   70.0  1.010  0.0  0.0       NaN    normal  notpresent  notpresent   
7    24.0    NaN  1.015  2.0  4.0    normal  abnormal  notpresent  notpresent   
8    52.0  100.0  1.015  3.0  0.0    normal  abnormal     present  notpresent   
9    53.0   90.0  1.020  2.0  0.0  abnormal  abnormal     present  notpresent   
10   50.0   60.0  1.010  2.0  4.0       NaN  abnormal     present  notpresent   
11   63.0   70.0  1.010  3.0  0.0  abnormal  abnormal     present  notpresent   
12   68.0   70.0  1.015  3.0  1.0       NaN    normal     present  notpresent   
13   68.0   70.0    NaN  NaN  NaN       NaN       NaN  notpresent  notpresent   
14   68.0   80.0  1.010  3.0  2.0    normal  abnormal     present     present   
15   40.0   80.0  1.015  3.0  0.0       NaN    normal  notpresent  notpresent   
16   47.0   70.0  1.015  2.0  0.0       NaN    normal  notpresent  notpresent   
17   47.0   80.0    NaN  NaN  NaN       NaN       NaN  notpresent  notpresent   
18   60.0  100.0  1.025  0.0  3.0       NaN    normal  notpresent  notpresent   
19   62.0   60.0  1.015  1.0  0.0       NaN  abnormal     present  notpresent   
20   61.0   80.0  1.015  2.0  0.0  abnormal  abnormal  notpresent  notpresent   
21   60.0   90.0    NaN  NaN  NaN       NaN       NaN  notpresent  notpresent   
22   48.0   80.0  1.025  4.0  0.0    normal  abnormal  notpresent  notpresent   
23   21.0   70.0  1.010  0.0  0.0       NaN    normal  notpresent  notpresent   
24   42.0  100.0  1.015  4.0  0.0    normal  abnormal  notpresent     present   
25   61.0   60.0  1.025  0.0  0.0       NaN    normal  notpresent  notpresent   
26   75.0   80.0  1.015  0.0  0.0       NaN    normal  notpresent  notpresent   
27   69.0   70.0  1.010  3.0  4.0    normal  abnormal  notpresent  notpresent   
28   75.0   70.0    NaN  1.0  3.0       NaN       NaN  notpresent  notpresent   
29   68.0   70.0  1.005  1.0  0.0  abnormal  abnormal     present  notpresent   
30    NaN   70.0    NaN  NaN  NaN       NaN       NaN  notpresent  notpresent   
31   73.0   90.0  1.015  3.0  0.0       NaN  abnormal     present  notpresent   
32   61.0   90.0  1.010  1.0  1.0       NaN    normal  notpresent  notpresent   
33   60.0  100.0  1.020  2.0  0.0  abnormal  abnormal  notpresent  notpresent   
34   70.0   70.0  1.010  1.0  0.0    normal       NaN     present     present   
35   65.0   90.0  1.020  2.0  1.0  abnormal    normal  notpresent  notpresent   
36   76.0   70.0  1.015  1.0  0.0    normal    normal  notpresent  notpresent   
37   72.0   80.0    NaN  NaN  NaN       NaN       NaN  notpresent  notpresent   
38   69.0   80.0  1.020  3.0  0.0  abnormal    normal  notpresent  notpresent   
39   82.0   80.0  1.010  2.0  2.0    normal       NaN  notpresent  notpresent   
40   46.0   90.0  1.010  2.0  0.0    normal  abnormal  notpresent  notpresent   
41   45.0   70.0  1.010  0.0  0.0       NaN    normal  notpresent  notpresent   
42   47.0  100.0  1.010  0.0  0.0       NaN    normal  notpresent  notpresent   
43   35.0   80.0  1.010  1.0  0.0  abnormal       NaN  notpresent  notpresent   
44   54.0   80.0  1.010  3.0  0.0  abnormal  abnormal  notpresent  notpresent   
45   54.0   80.0  1.020  3.0  0.0       NaN  abnormal  notpresent  notpresent   
46   48.0   70.0  1.015  0.0  0.0       NaN    normal  notpresent  notpresent   
47   11.0   80.0  1.010  3.0  0.0       NaN    normal  notpresent  notpresent   
48   73.0   70.0  1.005  0.0  0.0    

In [6]:
df.isnull().sum()

age        9
bp        12
sg        47
al        46
su        49
rbc      152
pc        65
pcc        4
ba         4
bgr       44
bu        19
sc        17
sod       87
pot       88
hemo      52
pcv       71
wbcc     106
rbcc     131
htn        2
dm         2
cad        2
appet      1
pe         1
ane        1
class      0
dtype: int64

### 4. Suppose that I dropped every row that contained at least one missing value. (In the context of analysis with missing data, we call this a "complete case analysis," because we keep only the complete cases!) How many rows would remain in our dataframe? What are at least two downsides to doing this?

> There's a good visual on slide 15 of [this deck](https://liberalarts.utexas.edu/prc/_files/cs/Missing-Data.pdf) that shows what a complete case analysis looks like if you're interested.

In [7]:
df_complete = df.dropna()

In [8]:
df_complete.shape

(158, 25)

Answer: There would be 158 rows.  This gives us significantly less data to work with than our original 400 rows of data, and we can never be sure these missing values were randomly missed - there may be bias in play that affects what data was missed which would translate to heavily biased data if we were to remove all rows with missing data.

### 5. Thinking critically about how our data were gathered, it's likely that these records were gathered by doctors and nurses. Brainstorm three potential areas (in addition to the missing data we've already discussed) where this data might be inaccurate or imprecise.

Answer: 
<br>
1) "Appetite" is a highly subjective thing to measure on a 1 to 5 ordinal scale.  
<br>
2) Missing data may be the result of improper record keeping or testing protocols
<br>

3) Differences in medical billing with insurance/state health coverage may give different patients differing levels of care, resulting in differing sets of data gathered and prognoses. 

---

## Step 4: Model the data.

### 6. Suppose that I want to construct a model where no person who has CKD will ever be told that they do not have CKD. What (very simple) model can I create that will never tell a person with CKD that they do not have CKD?

> Hint: Don't think about `statsmodels` or `scikit-learn` here.

Answer: 
Tell everyone who may potentially have CKD that they may be positive for CKD and should take further testing.  

### 7. In problem 6, what common classification metric did we optimize for? Did we minimize false positives or negatives?

Answer: Minimized false negatives - we optimized for sensitivity. 

### 8. Thinking ethically, what is at least one disadvantage to the model you described in problem 6?

Answer: Stressing individuals that likely do not have CKD.

Potential lawsuits with poor/bad faith choices in wording - Perhaps if this is determined to be done in bad faith as a medical billing scam?


### 9. Suppose that I want to construct a model where a person who does not have CKD will ever be told that they do have CKD. What (very simple) model can I create that will accomplish this?

Answer: Tell all individuals who have a chance that they do not have CKD, that they do not have CKD.

### 10. In problem 9, what common classification metric did we optimize for? Did we minimize false positives or negatives?

Answer: Minimized false positives - we optimized for specificity.

### 11. Thinking ethically, what is at least one disadvantage to the model you described in problem 9?

Answer: Individuals who have CKD will believe they do not, resultin in untreated disease leading to poor health outcomes/complications.

And, similarly, medical malpractice lawsuits.


### 12. Construct a logistic regression model in `sklearn` predicting class from the other variables. You may scale, select/drop, and engineer features as you wish - build a good model! Make sure, however, that you include at least one categorical/dummy feature and at least one quantitative feature.

> Hint: Remember to do a train/test split!

In [32]:
df.head()

age    bp     sg   al   su     rbc        pc         pcc          ba  \
0  48.0  80.0  1.020  1.0  0.0     NaN    normal  notpresent  notpresent   
1   7.0  50.0  1.020  4.0  0.0     NaN    normal  notpresent  notpresent   
2  62.0  80.0  1.010  2.0  3.0  normal    normal  notpresent  notpresent   
3  48.0  70.0  1.005  4.0  0.0  normal  abnormal     present  notpresent   
4  51.0  80.0  1.010  2.0  0.0  normal    normal  notpresent  notpresent   

     bgr    bu   sc    sod  pot  hemo   pcv    wbcc  rbcc  htn   dm cad appet  \
0  121.0  36.0  1.2    NaN  NaN  15.4  44.0  7800.0   5.2  yes  yes  no  good   
1    NaN  18.0  0.8    NaN  NaN  11.3  38.0  6000.0   NaN   no   no  no  good   
2  423.0  53.0  1.8    NaN  NaN   9.6  31.0  7500.0   NaN   no  yes  no  poor   
3  117.0  56.0  3.8  111.0  2.5  11.2  32.0  6700.0   3.9  yes   no  no  poor   
4  106.0  26.0  1.4    NaN  NaN  11.6  35.0  7300.0   4.6   no   no  no  good   

    pe  ane  class  
0   no   no      1  
1   no   no      1  
2   no  yes      1  
3  yes  yes      1  
4   no   no      1

In [33]:
df.replace({'notckd':0,'ckd':1}, inplace = True)

In [34]:
df.dropna(subset=['age', 'bp', 'bu', 'sc', 'htn', 'dm'], inplace = True)

In [35]:
X = df[['age', 'bp', 'bu', 'sc', 'htn', 'dm']]
y = df.iloc[:, -1]

In [36]:
X = pd.get_dummies(X, columns =['htn'])
X = pd.get_dummies(X, columns =['dm'])

In [37]:
X.head()

age    bp    bu   sc  htn_no  htn_yes  dm_no  dm_yes
0  48.0  80.0  36.0  1.2       0        1      0       1
1   7.0  50.0  18.0  0.8       1        0      1       0
2  62.0  80.0  53.0  1.8       1        0      0       1
3  48.0  70.0  56.0  3.8       0        1      1       0
4  51.0  80.0  26.0  1.4       1        0      1       0

In [38]:
X.drop(columns = 'htn_no', inplace = True)
X.drop(columns = 'dm_no', inplace = True)

In [39]:
X.head()

age    bp    bu   sc  htn_yes  dm_yes
0  48.0  80.0  36.0  1.2        1       1
1   7.0  50.0  18.0  0.8        0       0
2  62.0  80.0  53.0  1.8        0       1
3  48.0  70.0  56.0  3.8        1       0
4  51.0  80.0  26.0  1.4        0       0

In [40]:
X.shape

(359, 6)

In [41]:
y.head

<bound method NDFrame.head of 0      1
1      1
2      1
3      1
4      1
5      1
6      1
8      1
9      1
10     1
11     1
12     1
13     1
14     1
15     1
16     1
17     1
18     1
19     1
20     1
21     1
22     1
24     1
25     1
26     1
27     1
28     1
29     1
31     1
32     1
33     1
34     1
35     1
36     1
37     1
38     1
39     1
40     1
41     1
42     1
43     1
44     1
45     1
46     1
47     1
48     1
49     1
50     1
51     1
      ..
350    0
351    0
352    0
353    0
354    0
355    0
356    0
357    0
358    0
359    0
360    0
361    0
362    0
363    0
364    0
365    0
366    0
367    0
368    0
369    0
370    0
371    0
372    0
373    0
374    0
375    0
376    0
377    0
379    0
380    0
381    0
382    0
383    0
384    0
385    0
386    0
387    0
388    0
389    0
390    0
391    0
392    0
393    0
394    0
395    0
396    0
397    0
398    0
399    0
Name: class, Length: 359, dtype: int64>

In [42]:
y.shape

(359,)

In [43]:
X.head()

age    bp    bu   sc  htn_yes  dm_yes
0  48.0  80.0  36.0  1.2        1       1
1   7.0  50.0  18.0  0.8        0       0
2  62.0  80.0  53.0  1.8        0       1
3  48.0  70.0  56.0  3.8        1       0
4  51.0  80.0  26.0  1.4        0       0

In [44]:
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

%matplotlib inline
%config InlineBackend.figure_format = 'retina'

In [45]:
X_train, X_test, y_train, y_test = train_test_split(
                      X, y, test_size=0.2, random_state=42) # My X_test comes out as an array unless I specify test size. what's up with that?

In [46]:
#ss.fit(X_train)

ss = StandardScaler()
ss.fit(X_train)
X_train_sc = ss.transform(X_train)
X_test_sc = ss.transform(X_test)

/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype uint8, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype uint8, float64 were all converted to float64 by StandardScaler.
  """
/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: DataConversionWarning: Data with input dtype uint8, float64 were all converted to float64 by StandardScaler.
  


In [47]:
log_reg = LogisticRegression()

In [48]:
log_reg.fit(X_train_sc, y_train)

print(f'Logistic Regression Intercept: {log_reg.intercept_}')
print(f'Logistic Regression Coefficient: {log_reg.coef_}')

Logistic Regression Intercept: [2.39346137]
Logistic Regression Coefficient: [[-0.25559711  0.63736219  0.57478249  2.28771934  1.526365    1.60441518]]


/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [49]:
print(f'Log_reg predicted values: {log_reg.predict(X_test_sc)}')

Log_reg predicted values: [0 1 0 0 1 0 0 1 0 0 0 0 1 0 1 1 1 0 0 0 1 0 0 1 0 0 0 0 1 1 0 1 0 0 1 1 1
 0 1 1 1 1 0 1 1 0 1 1 1 0 0 0 1 1 0 0 1 1 1 0 1 1 0 1 1 1 0 1 0 0 1 1]


In [50]:
preds = log_reg.predict(X_train_sc)

In [51]:
np.exp(log_reg.coef_)

array([[0.77445392, 1.89148492, 1.77674403, 9.85244195, 4.60142023,
        4.9749493 ]])

In [52]:
np.exp(log_reg.coef_).mean()

3.978582393052614

In [53]:
log_reg.score(X_train_sc, y_train)

0.9024390243902439

In [54]:
log_reg.score(X_test_sc, y_test)

0.8888888888888888

---

## Step 5: Evaluate the model.

### 13. Based on your logistic regression model constructed in problem 12, interpret the coefficient of one of your quantitative features.

 A coefficient of 9.85 for the variable serum creatinine means there is a very strong positive correlation with elevated serum creatinine levels and chronic kidney disease.  

### 14. Based on your logistic regression model constructed in problem 12, interpret the coefficient of one of your categorical/dummy features.

 A coefficient of 4.6 for the dummied variable htn_yes, or hypertension, means there is a strong positive correlation with hypertension and chronic kidney disease. 

### 15. Despite being a relatively simple model, logistic regression is very widely used in the real world. Why do you think that's the case? Name at least two advantages to using logistic regression as a modeling technique.

Answer: 
<br>
1) It allows us to create coefficients for boolean values and other discrete variables
<br>
2) Regression can effectively occur on models with both discrete and continuous variables

### 16. Does it make sense to generate a confusion matrix on our training data or our test data? Why? Generate it on the proper data.

> Hint: Once you've generated your predicted $y$ values and you have your observed $y$ values, then it will be easy to [generate a confusion matrix using sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html).

Training data - This is the entire set of our data we are investigating.  

In [55]:
from sklearn.metrics import confusion_matrix

In [56]:
confusion_matrix(y_train, preds)

array([[112,   0],
       [ 28, 147]])

### 17. In this hospital case, we want to predict CKD. Do we want to optimize for sensitivity, specificity, or something else? Why? (If you don't think there's one clear answer, that's okay! There rarely is. Be sure to defend your conclusion!)

Answer: Sensitivity predominantly.  General accuracy is important too of course, but it's better for some people to be given false positives and be recommended for further screening, than to falsely tell people with chronic kidney disease that they have no medical issues, and allow their conditions to become worse untreated. 

### 18 (BONUS). Write a function that will create an ROC curve for you, then plot the ROC curve.

Here's a strategy you might consider:
1. In order to even begin, you'll need some fit model. Use your logistic regression model from problem 12.
2. We want to look at all values of your "threshold" - that is, anything where .predict() gives you above your threshold falls in the "positive class," and anything that is below your threshold falls in the "negative class." Start the threshold at 0.
3. At this value of your threshold, calculate the sensitivity and specificity. Store these values.
4. Increment your threshold by some "step." Maybe set your step to be 0.01, or even smaller.
5. At this value of your threshold, calculate the sensitivity and specificity. Store these values.
6. Repeat steps 3 and 4 until you get to the threshold of 1.
7. Plot the values of sensitivity and 1 - specificity.

### 19. Suppose you're speaking with the biostatistics lead at Mayo Clinic, who asks you "Why are unbalanced classes generally a problem? Are they a problem in this particular CKD analysis?" How would you respond?

Answer: Because when our dependent variable is unbalanced - when there is only a binary yes/no result and there is significantly more of one result than another, there may be a bias in the data that may would skew our results.  250 positives to 150 negatives isn't extremely imbalanced, but along with other factors generating bias, it may mean our sample isn't entirely reflective of the great population. 

### 20. Suppose you're speaking with a doctor at Mayo Clinic who, despite being very smart, doesn't know much about data science or statistics. How would you explain why unbalanced classes are generally a problem to this doctor?

Answer: They can lead to incorrect conclusions based off our limited data, causing some people to be misdiagonosed, or underdiagnosed, because we assume the information from our patients is representative of the population. 

### 21. Let's create very unbalanced classes just for the sake of this example! Generate very unbalanced classes by [bootstrapping](http://stattrek.com/statistics/dictionary.aspx?definition=sampling_with_replacement) (a.k.a. random sampling with replacement) the majority class.

1. The majority class are those individuals with CKD.
2. Generate a random sample of size 4,600 of individuals who have CKD **with replacement**. (Consider setting a random seed for this part!)
3. Create a new dataframe with the original data plus this random sample of data.
4. Now we should have a dataset with 5,000 observations, of which only about 150 are non-CKD individuals.

In [ ]:
np.random.int() #not sure how to do this...i could make several copies of the CKD individual datapoints and merge them but that would not be random. 

### 22. Build a logistic regression model on the unbalanced class data and evaluate its performance using whatever method(s) you see fit. How would you describe the impact of unbalanced classes on logistic regression as a classifier?

In [ ]:
X_train_unbal, X_test_unbal, y_train_unbal, y_test_unbal = train_test_split(
                      X, y, test_size=0.2, random_state=42) # My X_test comes out as an array unless I specify test size. what's up with that?
ss = StandardScaler()
ss.fit(X_train_unbal)
X_train_scunbal = ss.transform(X_train)
X_test_scunbal = ss.transform(X_test)


In [ ]:
log_reg = LogisticRegression

In [ ]:
log_reg.fit(X_train_scunbal, y_train_unbal)

print(f'Logistic Regression Intercept: {log_reg.intercept_}')
print(f'Logistic Regression Coefficient: {log_reg.coef_}')

In [ ]:
np.exp(log_reg.coef_)

In [ ]:
log_reg.score(X_train_scunbal, y_train_unbal)

In [ ]:
log_reg.score(X_test_scunbal, y_train_unbal)

In [ ]:
preds_unbal = log_reg.predict(X_train_scunbal)

In [ ]:
confusion_matrix(y_train_unbal, preds_unbal)

---

## Step 6: Answer the problem.

At this step, you would generally answer the problem! In this situation, you would likely present your model to doctors or administrators at the hospital and show how your model results in reduced false positives/false negatives. Next steps would be to find a way to roll this model and its conclusions out across the hospital so that the outcomes of patients with CKD (and without CKD!) can be improved!

Our model is presently around 90.2% accurate (259/287).  No false positives have been demonstrated between our training and testing data - Our specificity rate is 100% (147/147) which ensures no patients will be unnecessarily stressed with a false diagnoses.  However, our sensitivity rate is presently at 80% (112/140) which in a medical facility leaves much to be desired.  We must tune our model to achieve a higher sensitivity rate, as neglecting patients who are potentially affected by CKD is much more worrisome than having a few false positives that would necessitate further screening, and may serve as a warning for patients who do not have CKD but show symptoms that make them vulnerable. 